### ISSUE TITLE
Issue: Add threshold-based manual-pick metrics and configurable best validation checkpoint for EqTransformer training

### ISSUE THREAD

**Issue by ozw4** [ozw4/seis_hypo#321]
# Issue: Add threshold-based manual-pick metrics and configurable best validation checkpoint for EqTransformer training

## Background

The current EqTransformer training adapter can train on the phase-training station-window store and writes loss-based outputs such as:

```text
train_history.csv
valid_metrics.csv
test_metrics.csv
best_model.pth
last_model.pth
```

The current implementation saves `best_model.pth` when validation loss improves. This means **best-by-validation-loss checkpointing already exists**.

However, the current metrics are loss-only. It does not compute manual-pick detection metrics such as accuracy, precision, recall, or F1 at multiple probability thresholds.

We need validation/test metrics against manual P/S picks, evaluated at several thresholds, and optionally use one of those validation metrics to decide the best checkpoint.

## Goals

1. During training, evaluate P and S pick performance against manual pick labels on the validation split.
2. Report accuracy, precision, recall, and F1 for multiple thresholds.
3. Also evaluate the final/best model on the test split using the same metrics.
4. Preserve current `best_model.pth` behavior by default: save best checkpoint by validation loss.
5. Add an optional config setting to save `best_model.pth` by a selected validation metric, e.g. `valid/P/f1@0.50`.

## Existing behavior to preserve

The existing behavior should remain valid:

```text
best_model.pth
  saved when valid_loss improves

last_model.pth
  saved after the final epoch

valid_metrics.csv / test_metrics.csv
  still include loss
```

If no new metric-selection config is set, checkpoint selection must remain validation-loss based.

## Proposed config

Extend the EqT training YAML schema with an optional `evaluation` block.

Example:

```yaml
evaluation:
  enabled: true
  thresholds: [0.05, 0.10, 0.20, 0.30, 0.50]
  phases: ['P', 'S']
  pick_tolerance_samples: 50
  include_background: true
  save_best_by:
    metric: valid_loss
    mode: min
```

Alternative F1-based checkpointing:

```yaml
evaluation:
  enabled: true
  thresholds: [0.05, 0.10, 0.20, 0.30, 0.50]
  phases: ['P', 'S']
  pick_tolerance_samples: 50
  include_background: true
  save_best_by:
    metric: valid_P_f1_thr_0.30
    mode: max
```

Rules:

- `thresholds` must be non-empty values in `[0, 1]`.
- `phases` supports `P` and `S` initially.
- `pick_tolerance_samples` must be non-negative.
- `save_best_by.metric: valid_loss` with `mode: min` should be the default.
- If a metric name is unknown, raise a clear config/validation error.

## Metric definition

Evaluate per phase (`P`, `S`) and per threshold.

For each selected row in a split:

### Ground truth

Use the CSV arrival-sample columns, not just the target image:

```text
P -> p_arrival_sample
S -> s_arrival_sample
```

A row is ground-truth positive for a phase if:

```text
arrival sample is present and finite
```

Background rows are ground-truth negative for P and S.

### Prediction

Use the model output channel corresponding to the phase:

```text
output channel 1 -> P
output channel 2 -> S
```

For a given phase and threshold:

```text
score = max probability over samples
predicted_positive = score >= threshold
predicted_sample = argmax probability over samples
```

### Match criterion

A prediction is a true positive if:

```text
ground_truth_positive
and predicted_positive
and abs(predicted_sample - manual_pick_sample) <= pick_tolerance_samples
```

False positive:

```text
predicted_positive and not true_positive
```

False negative:

```text
ground_truth_positive and not true_positive
```

True negative:

```text
not ground_truth_positive and not predicted_positive
```

Then compute:

```text
accuracy  = (TP + TN) / (TP + FP + FN + TN)
precision = TP / (TP + FP)
recall    = TP / (TP + FN)
f1        = 2 * precision * recall / (precision + recall)
```

Use `0.0` for precision/recall/F1 when the denominator is zero, unless there is already a repo convention for NaN metrics.

## Output files

Add these files under the run directory:

```text
valid_pick_metrics.csv
test_pick_metrics.csv
```

Recommended columns:

```text
split
phase
threshold
pick_tolerance_samples
n_examples
n_positive
n_negative
tp
fp
fn
tn
accuracy
precision
recall
f1
mean_abs_error_samples
median_abs_error_samples
```

`mean_abs_error_samples` and `median_abs_error_samples` should be computed over true positives only.

Also add epoch-level validation metric history:

```text
valid_pick_metrics_by_epoch.csv
```

Recommended columns:

```text
epoch
phase
threshold
tp
fp
fn
tn
accuracy
precision
recall
f1
```

This allows plotting F1 vs epoch later.

## Training loop changes

Likely file:

```text
src/phase_training/adapters/seisbench_eqt.py
```

Required changes:

1. Add an `EqtEvaluationConfig` dataclass.
2. Parse `evaluation:` in `load_eqt_train_config()`.
3. Extend `StationWindowEqtDataset.__getitem__()` or row handling so evaluation can access:
   - `p_arrival_sample`
   - `s_arrival_sample`
   - `example_type`
   - `example_id`
4. Add an evaluation function such as:

```python
def evaluate_pick_metrics(
    model,
    loader,
    *,
    device,
    thresholds,
    phases,
    pick_tolerance_samples,
) -> list[dict[str, Any]]:
    ...
```

5. At the end of each epoch:
   - compute `valid_loss`
   - compute validation pick metrics if enabled
   - write/append epoch metrics
   - decide whether to update `best_model.pth`
6. After training:
   - load `best_model.pth` before final validation/test metric reporting, or clearly document whether final metrics are from last model vs best model. Preferred behavior: final valid/test metrics are from `best_model.pth`.

## Best checkpoint selection

Current behavior:

```text
best_model.pth is saved by minimum validation loss
```

Required behavior:

- Preserve current default.
- Add optional selection by validation pick metric.

Examples:

```yaml
evaluation:
  save_best_by:
    metric: valid_loss
    mode: min
```

or:

```yaml
evaluation:
  save_best_by:
    metric: valid_P_f1_thr_0.30
    mode: max
```

The metric naming must be deterministic and documented. For example, threshold `0.30` can become:

```text
valid_P_f1_thr_0.30
valid_S_recall_thr_0.50
```

If `save_best_by.metric` references a pick metric, checkpoint selection should use that metric after each validation evaluation.

Also write a metadata JSON file:

```text
best_model_metadata.json
```

Suggested fields:

```json
{
  "epoch": 7,
  "selection_metric": "valid_P_f1_thr_0.30",
  "selection_mode": "max",
  "selection_value": 0.82,
  "valid_loss": 0.123,
  "checkpoint": "best_model.pth"
}
```

## Performance considerations

Pick-metric evaluation requires full model inference on the validation split each epoch. This is acceptable for now, but implement it without gradients:

```python
model.eval()
with torch.no_grad():
    ...
```

If runtime is too high, later add:

```yaml
evaluation:
  every_n_epochs: 1
```

Do not add this unless needed for the initial implementation.

## Tests

Add tests around the metric computation without needing a real EqTransformer model.

Suggested tests:

### Pure metric tests

Create a small pure function test with synthetic arrays:

- one true P pick at sample 100
- prediction peak at sample 105
- tolerance 10 -> TP
- tolerance 3 -> FP + FN
- background row with peak above threshold -> FP
- background row below threshold -> TN

Assert TP/FP/FN/TN and accuracy/precision/recall/F1.

### Dataset row tests

Verify evaluation has access to:

```text
p_arrival_sample
s_arrival_sample
example_type
```

### Config tests

- `evaluation.thresholds` parses.
- invalid threshold raises.
- `save_best_by.metric=valid_loss` works.
- invalid `save_best_by.mode` raises.

### Training integration smoke test

Use a tiny fake dataset/model or monkeypatch model output to verify:

- `valid_pick_metrics.csv` is written.
- `test_pick_metrics.csv` is written.
- `valid_pick_metrics_by_epoch.csv` is written.
- `best_model_metadata.json` is written.
- `best_model.pth` updates according to the configured metric.

## Acceptance criteria

- Existing training without an `evaluation:` block still works and saves `best_model.pth` by minimum validation loss.
- With `evaluation.enabled: true`, validation and test pick metrics are written for every configured threshold and phase.
- Metrics include accuracy, precision, recall, F1, TP, FP, FN, TN.
- Manual pick matching uses `p_arrival_sample` / `s_arrival_sample` from the examples CSV.
- Background examples contribute negatives for P/S.
- `best_model.pth` can still be selected by `valid_loss`.
- `best_model.pth` can optionally be selected by a configured validation F1/precision/recall metric.
- The run directory contains enough metadata to know which epoch and metric produced `best_model.pth`.
